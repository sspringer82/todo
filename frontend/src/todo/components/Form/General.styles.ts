import styled from 'styled-components';
import {
  Select as MUISelect,
  TextField as MUITextField,
  Divider as MUIDivider,
} from '@material-ui/core';
import { DateTimePicker as MUIDateTimePicker } from '@material-ui/pickers';

export const Select = styled(MUISelect)`
  width: 100%;
`;

export const TextField = styled(MUITextField)`
  width: 100%;
`;

export const DateTimePicker = styled(MUIDateTimePicker)`
  width: 100%;
`;

export const Divider = styled(MUIDivider)`
  && {
    margin: 10px 0;
  }
`;
