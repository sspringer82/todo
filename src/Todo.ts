export type Todo = {
    id: number,
    title: string,
    done: boolean,
    category?: string,
    color?: string,
    comment?: string,
    subtask?: Subtask[]
};

export type TodoInput = {id?: number} & Omit<Todo, 'id'>;

export type Subtask = {todoId: number} & Omit<Todo, 'comment' | 'subtask'>

export type SubtaskInput = {id?: number} & Omit<Subtask, 'id'>

export const initialTodo:TodoInput = {
    title: '',
    done: false,
    comment: ''
};
Object.freeze(initialTodo);