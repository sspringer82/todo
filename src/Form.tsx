import React, { ChangeEvent, FormEvent, useState } from "react";
import { TodoInput } from "./Todo";
import produce from "immer";
import useTodoService from "./useTodoService";

const initialTodo: TodoInput = {
  title: "",
  done: false,
};

const Form: React.FC = () => {
  const [todo, setTodo] = useState<TodoInput>(initialTodo);
  const {save} = useTodoService();

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const value = e.target.value;
    const name = e.target.name as keyof TodoInput;
    setTodo((prevState) =>
      produce(prevState, (draftState) => {
        draftState[name] = value as never;
      })
    );
  }

  async function handleSubmit(e: FormEvent) {
    e.preventDefault();
    await save(todo);
    setTodo(initialTodo);
  }

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        onChange={handleChange}
        value={todo.title}
        name="title"
      />
      <button type="submit">save</button>
    </form>
  );
};

export default Form;
