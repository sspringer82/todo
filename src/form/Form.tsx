import React, { FormEvent } from "react";
import Subtask from "../Subtask/Subtask";
import { Todo, TodoInput } from "../Todo";
import useForm from "../useForm";

export type Props = {
  todo: TodoInput,
  onSubmit: (e: FormEvent) => Promise<void>,
}

const Form: React.FC<Props> = ({todo, onSubmit}) => {
  const { handleChange, item } = useForm<TodoInput>(
    todo,
  );

  return (
    <>
      <form
        onSubmit={onSubmit}
        autoComplete="off"
      >
        <fieldset>
          <label>
            Title:{" "}
            <input
              type="text"
              name="title"
              value={item.title}
              onChange={handleChange}
            />
          </label>
        </fieldset>
        <fieldset>
          <label>
            Done:{" "}
            <input
              type="checkbox"
              name="done"
              checked={item.done}
              onChange={handleChange}
            />
          </label>
        </fieldset>
        <fieldset>
          <label>
            Comment:
            <textarea
              name="comment"
              value={item.comment}
              onChange={handleChange}
            ></textarea>
          </label>
        </fieldset>
        <button type="submit">save</button>
      </form>
      {item.id && <Subtask todo={item as Todo} />}
    </>
  );
};

export default Form;
