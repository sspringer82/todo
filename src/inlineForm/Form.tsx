import React, { useEffect } from 'react';
import { initialTodo, Todo, TodoInput } from '../Todo';
import useForm from '../useForm';

export type Props = {
  todo?: Todo;
  onSave: (item: TodoInput) => Promise<void>;
  onCancel: () => void;
};

// @todo generic component?
const Form: React.FC<Props> = ({ todo, onSave, onCancel }) => {
  const { handleSubmit, handleChange, item, setItem } = useForm<TodoInput>(
    initialTodo,
    onSave,
  );
  useEffect(() => {
    if (todo) {
      setItem(todo);
    }
  }, [todo, setItem]);

  return (
    <form onSubmit={handleSubmit} data-testid="inlineForm">
      <input
        type="text"
        onChange={handleChange}
        value={item.title}
        name="title"
        data-testid="title-input"
      />
      <button type="submit" data-testid="submit-button">
        save
      </button>
      <button
        type="button"
        onClick={() => onCancel()}
        data-testid="cancel-button"
      >
        cancel
      </button>
    </form>
  );
};

export default Form;
