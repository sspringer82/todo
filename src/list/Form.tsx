import React from 'react';
import { TodoInput } from '../Todo';
import useForm from '../useForm';
import useTodoService from '../useTodoService';

const initialTodo: TodoInput = {
  title: '',
  done: false,
};

const Form: React.FC = () => {
  const {save} = useTodoService();
  const {handleSubmit, handleChange, item} = useForm<TodoInput>(initialTodo, save);

  return (
    <form onSubmit={handleSubmit}>
      <input
        type="text"
        onChange={handleChange}
        value={item.title}
        name="title"
      />
      <button type="submit">save</button>
    </form>
  );
};

export default Form;
