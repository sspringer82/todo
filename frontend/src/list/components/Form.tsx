import React, {
  ChangeEvent,
  FormEvent,
  useState,
  useEffect,
  useRef,
} from 'react';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Button,
  IconButton,
  Select,
  InputLabel,
  Input,
  MenuItem,
} from '@material-ui/core';
import { useHistory, useParams } from 'react-router';
import update from 'immutability-helper';
import { InputTypeList, List } from '../../shared/List';
import { useSelector, useDispatch } from 'react-redux';
import { AppState } from '../../reducers/rootReducer';
import DeleteIcon from '@material-ui/icons/Delete';
import { deleteListAction, saveListAction } from '../actions/list.actions';
import { getUsers } from '../../user/selectors/user.selector';
import { loadUsersAction } from '../../user/actions/user.actions';
import { getSettings } from '../../settings/selectors/settings.selector';
import { saveSettingsAction } from '../../settings/actions/settings.actions';
import { TextField, FormControl } from './Form.styles';

const Form: React.FC = () => {
  const params = useParams<{ id: string }>();
  const history = useHistory();
  const dispatch = useDispatch();
  const users = useSelector(getUsers);
  const listName = useRef<HTMLInputElement>(null);

  let initialList: InputTypeList = {
    name: '',
    sharedWith: [],
  };

  const foundList = useSelector((state: AppState) =>
    state.list.lists.find((list: List) => list.id === parseInt(params.id, 10))
  );

  if (foundList) {
    initialList = foundList as InputTypeList;
  }

  const [list, setList] = useState<InputTypeList>(initialList);

  function handleClose() {
    history.push('/');
  }

  function handleChange(e: ChangeEvent<HTMLInputElement>) {
    const field = e.currentTarget.name;
    const value = e.currentTarget.value;

    setList((prevTodo: InputTypeList) =>
      update(prevTodo, { [field]: { $set: value } })
    );
  }

  function handleShared(
    e: ChangeEvent<{ name?: string | undefined; value: unknown }>
  ) {
    const data = users.filter(user =>
      ((e.target.value as unknown) as number[])!.includes(user.id)
    );
    setList((prevTodo: InputTypeList) =>
      update(prevTodo, { sharedWith: { $set: data } })
    );
  }

  useEffect(() => {
    setTimeout(() => {
      listName.current && listName.current!.focus();
    }, 0);
  }, []);

  useEffect(() => {
    dispatch(loadUsersAction.request());
  }, [dispatch]);

  function handleSubmit(e: FormEvent<HTMLFormElement>) {
    e.preventDefault();
    dispatch(saveListAction.request(list));
    handleClose();
  }

  const settings = useSelector(getSettings);
  function handleDelete() {
    dispatch(deleteListAction.request(list as List));
    dispatch(
      saveSettingsAction.request(
        update(settings, { list: { $set: undefined } })
      )
    );
    handleClose();
  }

  return (
    <Dialog onClose={handleClose} open={true}>
      <form onSubmit={handleSubmit}>
        <DialogTitle>Liste anlegen/editieren</DialogTitle>
        <DialogContent>
          <div>
            <TextField
              label="Liste"
              name="name"
              onChange={handleChange}
              value={list.name}
              inputRef={listName}
            />
          </div>
          <FormControl>
            <InputLabel id="demo-mutiple-name-label">Name</InputLabel>
            <Select
              labelId="demo-mutiple-name-label"
              id="demo-mutiple-name"
              name="sharedWith"
              multiple
              value={
                (list.sharedWith && list.sharedWith!.map(user => user.id)) || []
              }
              onChange={handleShared}
              input={<Input />}
            >
              {users.map(user => (
                <MenuItem key={user.id} value={user.id}>
                  {user.username}
                </MenuItem>
              ))}
            </Select>
          </FormControl>
        </DialogContent>
        <DialogActions>
          {list.id && (
            <IconButton onClick={handleDelete}>
              <DeleteIcon />
            </IconButton>
          )}
          <Button type="submit">speichern</Button>
          <Button color="secondary" onClick={handleClose}>
            abbrechen
          </Button>
        </DialogActions>
      </form>
    </Dialog>
  );
};

export default Form;
