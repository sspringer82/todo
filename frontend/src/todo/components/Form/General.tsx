import React, { ChangeEvent } from 'react';
import {
  Typography,
  FormControlLabel,
  Switch,
  InputLabel,
  MenuItem,
} from '@material-ui/core';

import moment, { Moment } from 'moment';
import { useSelector } from 'react-redux';
import { getLists } from '../../../list/selectors/list.selector';
import { InputTypeTodo } from '../../../shared/Todo';
import { Select, TextField, DateTimePicker, Divider } from './General.styles';

interface Props {
  todo: InputTypeTodo;
  tabIndex: number;
  tab: number;
  handleChange: (
    e:
      | ChangeEvent<HTMLInputElement>
      | {
          currentTarget: {
            name: string;
            value: string;
          };
        },
  ) => void;
}

const General: React.FC<Props> = ({ todo, tabIndex, tab, handleChange }) => {
  const lists = useSelector(getLists);
  return (
    <Typography
      component="div"
      role="tabpanel"
      hidden={tab !== tabIndex}
      id={`simple-tabpanel-${tabIndex}`}
      aria-labelledby={`simple-tab-${tabIndex}`}
    >
      <div>
        <TextField
          label="Aufgabe"
          name="title"
          onChange={handleChange}
          value={todo.title}
        />
      </div>
      <div>
        <FormControlLabel
          control={
            <Switch
              checked={todo.done}
              onChange={handleChange}
              value={true}
              name="done"
            />
          }
          label="Erledigt"
        />
      </div>
      <Divider />
      <DateTimePicker
        clearable
        value={todo.due}
        onChange={(date: Moment | null) =>
          handleChange(({
            currentTarget: {
              name: 'due',
              value: date ? date.format() : null,
            },
          } as unknown) as ChangeEvent<HTMLInputElement>)
        }
        name="due"
        label="Zu erledigen bis"
      />
      <Divider />
      <div>
        <InputLabel id="list-label">Liste</InputLabel>
        <Select
          labelId="list-label"
          value={todo.list && todo.list.id ? todo.list.id : ''}
          name="list"
          onChange={(e: any) => {
            handleChange({
              currentTarget: {
                name: e.target.name as string,
                value: (lists.find(
                  (list) => list.id === parseInt(e.target.value as string, 10),
                ) as unknown) as string,
              },
            });
          }}
        >
          {lists.map((list) => (
            <MenuItem value={list.id} key={list.id}>
              {list.name}
            </MenuItem>
          ))}
        </Select>
      </div>
      <Divider />
      <div>
        Erstellt am: {moment(todo.createdAt).format('DD.MM.YYYY hh:mm')}
      </div>
    </Typography>
  );
};

export default General;
