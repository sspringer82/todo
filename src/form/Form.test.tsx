import { act, fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import { Todo } from "../Todo";
import Form from "./Form.container";

describe("Form", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  it("should perform a snapshot test", () => {
    const { container } = render(
      <MemoryRouter>
        <Form></Form>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render an empty form", async () => {
    act(() => {
      render(
        <MemoryRouter>
          <Form></Form>
        </MemoryRouter>
      );
    });
    expect(screen.getByTestId("title")).toHaveValue("");
    expect(screen.getByTestId("done")).not.toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue("");
  });
  it("should load an existing entry into the form", async () => {
    (fetch as any).mockResponse(
      JSON.stringify([
        {
          id: 1,
          title: "first todo",
          done: true,
          comment: "comment",
          category: { id: 1, title: "Private", color: "green" },
          due: "2022-06-01T08:15:00",
          created: "2021-06-01T08:15:00",
        },
      ])
    );
    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/edit/1"]}>
          <Route path="/edit/:id">
            <Form></Form>
          </Route>
        </MemoryRouter>
      );
    });
    expect(screen.getByTestId("title")).toHaveValue("first todo");
    expect(screen.getByTestId("done")).toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue("comment");
    expect(screen.getByTestId("due")).toHaveValue("2022-06-01T08:15:00");
  });
  it("should save a new entry", async () => {
    await act(async () =>
      render(
        <MemoryRouter>
          <Form></Form>
        </MemoryRouter>
      )
    );
    act(() => {
      fireEvent.change(screen.getByTestId("title"), {
        target: { value: "New Title" },
      });
      fireEvent.click(screen.getByTestId("done"));
      fireEvent.change(screen.getByTestId("comment"), {
        target: { value: "New Comment" },
      });
      fireEvent.change(screen.getByTestId("due"), {
        target: { value: "2022-06-01T08:15:00" },
      });
    });
    expect(screen.getByTestId("title")).toHaveValue("New Title");
    expect(screen.getByTestId("done")).toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue("New Comment");
    (fetch as any).mockResponse(
      JSON.stringify([
        {
          id: 1,
          title: "New Title",
          done: true,
          comment: "New Comment",
          category: { id: 1, title: "Private", color: "green" },
          due: "2022-06-01T08:15:00",
          created: "2021-06-01T08:15:00",
        },
      ])
    );
    await act(async () => {
      fireEvent.click(screen.getByTestId("submit"));
    });
    expect(fetchMock.mock.calls[0][1].method).toBe("POST");
    expect(fetchMock.mock.calls[0][1].headers["Content-Type"]).toBe(
      "Application/json"
    );
    const body = JSON.parse(fetchMock.mock.calls[0][1].body);
    expect(body.title).toBe("New Title");
    expect(body.done).toBe(true);
    expect(body.comment).toBe("New Comment");
  });
  it("should save an existing entry", async () => {
    (fetch as any).mockResponseOnce(
      JSON.stringify([
        {
          id: 2,
          title: "Existing Todo",
          done: true,
          comment: "Existing Comment",
          category: { id: 1, title: "Private", color: "green" },
          due: "2022-06-01T08:15:00",
          created: "2021-06-01T08:15:00",
        },
      ])
    );
    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/edit/2"]}>
          <Route path="/edit/:id">
            <Form></Form>
          </Route>
        </MemoryRouter>
      );
    });
    expect(screen.getByTestId("title")).toHaveValue("Existing Todo");
    expect(screen.getByTestId("done")).toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue("Existing Comment");
    expect(screen.getByTestId("due")).toHaveValue("2022-06-01T08:15:00");
    act(() => {
      fireEvent.change(screen.getByTestId("title"), {
        target: { value: "Existing Todo Modified" },
      });
      fireEvent.click(screen.getByTestId("done"));
      fireEvent.change(screen.getByTestId("comment"), {
        target: { value: "Existing Comment Modified" },
      });
      fireEvent.change(screen.getByTestId("due"), {
        target: { value: "2022-08-01T08:15:00" },
      });
    });
    expect(screen.getByTestId("title")).toHaveValue("Existing Todo Modified");
    expect(screen.getByTestId("done")).not.toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue(
      "Existing Comment Modified"
    );
    expect(screen.getByTestId("due")).toHaveValue("2022-08-01T08:15:00");
    (fetch as any).mockResponseOnce(
      JSON.stringify([
        {
          id: 2,
          title: "Existing Todo Modified",
          done: false,
          comment: "Existing Comment Modified",
          category: { id: 1, title: "Private", color: "green" },
          due: "2022-08-01T08:15:00",
          created: "2021-06-01T08:15:00",
        },
      ])
    );
    await act(async () => {
      fireEvent.click(screen.getByTestId("submit"));
    });
    expect(fetchMock.mock.calls[1][1].method).toBe("PUT");
    expect(fetchMock.mock.calls[1][1].headers["Content-Type"]).toBe(
      "Application/json"
    );
    const body = JSON.parse(fetchMock.mock.calls[1][1].body);
    expect(body.title).toBe("Existing Todo Modified");
    expect(body.done).toBe(false);
    expect(body.comment).toBe("Existing Comment Modified");
    expect(body.due).toBe("2022-08-01T08:15:00");
  });
  it("should show a list of subtasks", async () => {
    const response: Todo[] = [
      {
        id: 1,
        title: "Get up",
        done: true,
        subtasks: [],
        category: { id: 1, title: "Private", color: "green" },
        due: "",
        created: "2021-06-01T08:15:00",
      },
      {
        id: 2,
        title: "Eat",
        done: true,
        category: { id: 1, title: "Private", color: "green" },
        subtasks: [
          {
            id: 1,
            title: "sit down",
            todoId: 1,
            done: false,
            created: "2021-06-01T08:15:00",
          },
          {
            id: 2,
            title: "open your mouth",
            todoId: 1,
            done: false,
            created: "2021-06-01T08:15:00",
          },
        ],
        due: "",
        created: "2021-06-01T08:15:00",
      },
      {
        id: 3,
        title: "Sleep",
        done: false,
        subtasks: [],
        category: { id: 1, title: "Private", color: "green" },
        due: "",
        created: "2021-06-01T08:15:00",
      },
    ];

    (fetch as any).mockResponse(JSON.stringify(response));
    await act(async () => {
      render(
        <MemoryRouter initialEntries={["/edit/2"]}>
          <Route path="/edit/:id">
            <Form></Form>
          </Route>
        </MemoryRouter>
      );
    });
    expect(screen.getAllByTestId("title")[0]).toHaveValue("Eat");
    expect(screen.getByTestId("done")).toBeChecked();
    expect(screen.getByTestId("comment")).toHaveValue("");

    expect(screen.getAllByTestId("listItem-container")).toHaveLength(2);
    expect(screen.getAllByTestId("title")[1]).toHaveTextContent("sit down");
    expect(screen.getAllByTestId("title")[2]).toHaveTextContent(
      "open your mouth"
    );
  });
});
