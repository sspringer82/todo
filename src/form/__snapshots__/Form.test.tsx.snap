// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should perform a snapshot test 1`] = `
<div>
  <form
    autocomplete="off"
    class="flex flex-col gap-4"
  >
    <div
      class="border relative p-1"
      style="border-color: #34d399;"
    >
      <label
        class="absolute bg-gray-800 text-xs px-1"
        style="top: -12px; left: 0px; color: rgb(52, 211, 153);"
      >
        Title
      </label>
      <input
        class="bg-gray-900 py-1 pl-1 focus:outline-none"
        data-testid="title"
        label="Title"
        name="title"
        style="width: 100%;"
        type="text"
        value=""
      />
    </div>
    <div
      class="border relative p-1"
      style="border-color: #34d399;"
    >
      <label
        class="absolute bg-gray-800 text-xs px-1"
        style="top: -12px; left: 0px; color: rgb(52, 211, 153);"
      >
        Done
      </label>
      <div
        aria-checked="false"
        data-testid="done"
        role="checkbox"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          style="color: red;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </div>
    </div>
    <div
      class="border relative p-1"
      style="border-color: #34d399;"
    >
      <label
        class="absolute bg-gray-800 text-xs px-1"
        style="top: -12px; left: 0px; color: rgb(52, 211, 153);"
      >
        Comment
      </label>
      <textarea
        class="bg-gray-900 py-1 pl-1 focus:outline-none"
        data-testid="comment"
        label="Comment"
        name="comment"
        style="margin-bottom: -7px; width: 100%;"
      />
    </div>
    <div>
      <button
        class="px-5 py-1 bg-opacity-25 focus:outline-none border"
        data-testid="submit"
        style="border-color: #34d399; background-color: rgba(52, 211, 153, 0.25);"
        type="submit"
        variant="primary"
      >
        save
      </button>
    </div>
  </form>
  <div
    class="flex"
    style="margin: 10px -10px 15px -10px;"
  >
    <div
      style="height: 5px; width: 5px; background-color: rgb(52, 211, 153);"
    />
    <div
      style="margin-top: 4px; height: 0px; border-top: 1px solid #34D399; flex-grow: 1;"
    />
    <div
      style="margin-top: 4px; margin-left: 7px; margin-right: 10px; height: 8px; width: 16px; background-color: rgb(52, 211, 153); transform: skew(70deg);"
    />
    <div
      style="height: 0px; border-top: 1px solid #34D399; flex-grow: 2; margin-top: 11px;"
    />
    <div
      style="margin-top: 4px; margin-left: 7px; margin-right: 10px; height: 8px; width: 16px; background-color: rgb(52, 211, 153); transform: skew(-70deg);"
    />
    <div
      style="height: 0px; border-top: 1px solid #34D399; flex-grow: 1; margin-top: 4px;"
    />
    <div
      style="height: 5px; width: 5px; background-color: rgb(52, 211, 153);"
    />
  </div>
  <button
    class="px-5 py-1 bg-opacity-25 focus:outline-none border"
    style="border-color: #34d399; margin-top: 10px;"
  >
    close
  </button>
</div>
`;
